designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,75 @@
+- Start Date: 2019-09-28
+- RFC PR: https://github.com/eslint/rfcs/pull/40
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# Moving to Asynchronous API
+
+## Summary
+
+This RFC adds a new class `ESLint` that has asynchronous API and deprecates `CLIEngine`.
+
+## Motivation
+
+- Dynamic `import()` has arrived at Stage 4. The dynamic loading of ES modules requires asynchronous API. Migrating to asynchronous API opens up doors to write plugins/configs with ES modules.
+- Linting in parallel requires asynchronous API. We can improve linting logic to run it in parallel. And we can improve config loading logic and file enumeration logic to run it in parallel. (E.g., load `extends` list, traverse child directories.)
+
+Because the migration of public API needs long time, we should start to migrate our APIs earlier.
+
+And the name of `CLIEngine`, our primary API, causes confusion to the community. People try `Linter` class at first, then they notice it doesn't work as expected. We have a lot of issues that say "please use `CLIEngine` instead."
+The name of new class, `ESLint`, is our primary API clearly.
+
+## Detailed Design
+
+### Add new `ESLint` class
+
+This RFC adds a new class `ESLint`. It has almost the same methods as `CLIEngine`, but the following methods return `Promise`.
+
+- `executeOnFiles()`
+- `executeOnText()`
+- `getConfigForFile()`
```

Comment:
Why do all of the methods below need to be async? Other then non-blocking IO, what's the advantages of changing them to async? `getConfigForFile` for example, should always execute on the main thread, so we could take full advantage of caching. `executeOnFiles` should be async - yes. I'm not even sure that `executeOnText` should be async, because I don't know what would be the advantage. But the rest of them should stay sync, in my opinion.

Reply:
Because it's impossible to import ES modules synchronously. Probably Node.js gets stable ES modules in this month (v12.12.0). This is the good timing to change the public API to async rather than the future breaking change.

- `executeOnFiles()` depends on module system and file system
- `executeOnText()` depends on module system and file system to load configs
- `getConfigForFile()` depends on module system and file system
- `getFormatter()` depends on module system to load custom formatters.
- `isPathIgnored()` depends on module system and file system to load configs
- `outputFixes()` depends on file system.

Reply:
Just to clarify, when we are talking about importing ES modules, are we talking about our code importing Node native modules like `fs`, `path`, etc? Us importing our own modules like `options`, `log`, etc. Or plugin authors importing public ESLint modules like `linter` and `CLIEngine`?

Reply:
No. We are talking about the community use ESLint along with ES modules. For details:

- End-users use ESLint to develop an application that is `type: module`. In this case, the users cannot use `.eslintrc.js` because `require()` cannot load it. We can resolve the issue if the config loader uses `import()` syntax instead of `require()`.
- Plugin/Shareable-config authors publish plugin/config packages that are `type: module`. In this case, end-users cannot use the package because `require()` cannot load it. We can resolve the issue if the config loader uses `import()` syntax instead of `require()`.

Then `import()` syntax returns `Promise`. We cannot use it with sync APIs.

Reply:
OK. Thanks for the clarification. I would love to hear from other members of the team. I know I'm minority, but I do not usually chase after every new and shiny thing. For me, CommonJS works perfectly fine and I would not consider either of the scenarios described above as critical in any way. But as I said above, my position is probably not very common. I know @nzakas has not been able to spend much time online lately, but would love to hear his opinion on this RFC as well.

Reply:
Well, it may enforce ESLint users to block the use of ES modules because end-users cannot use ESLint with `.eslintrc.js` when they develop an application that is ES modules. It's not healthy for the JS community in my 2 cents.